advection_test/functions.py

```diff
@@ -0,0 +1,1302 @@
+def build_tracerAdvection(configuration, fvt_dict, tracers):
+    """
+    Use: tracAdv = build_tracerAdvection(configuration, fvt_dict, tracers)
+
+    Initializes the tracer advection class from FiniteVolumeTransport
+    and TracerAdvection.
+
+    Inputs:
+    - configuration: Dict{'stencil_factory', 'grid_data',
+                          'damping_coefficients'}
+    - fvt_dict: Dict{'grid_type', 'hord'}
+    - tracers: Dict{'smoke}
+
+    Outputs:
+    - tracAdv: class instance that performs tracer advection
+    """
+
+    fvtp_2d = FiniteVolumeTransport(
+        configuration["stencil_factory"],
+        configuration["grid_data"],
+        configuration["damping_coefficients"],
+        fvt_dict["grid_type"],
+        fvt_dict["hord"],
+    )
+
+    tracAdv = TracerAdvection(
+        configuration["stencil_factory"],
+        fvtp_2d,
+        configuration["grid_data"],
+        configuration["communicator"],
+        tracers,
+    )
+
+    return tracAdv
+
+
+def prepare_everythingForAdvection(
+    configuration,
+    uc,
+    vc,
+    delp,
+    tracers,
+    density,
+    dimensions,
+    units,
+    origins,
+    backend,
+    fvt_dict,
+    dt_acoustic=300,
+):
+    """
+    Use: fluxPrep, tracAdv = prepare_everythingForAdvection(
+        configuration, uc, vc, delp, tracers, density, dimensions, units,
+        origins, backend, fvt_dict, dt_acoustic=300)
+
+    Inputs:
+    - configuration: Dict{'stencil_factory', 'grid_data', 'damping_coefficients'}
+    - uc, vc: x- and y- winds on C-grid
+    - delp: pressure thickness of layer
+    - tracers: Dict{'smoke}
+    - density: (kg/m3 for air)
+    - dimensions: Dict{'nx', 'ny', 'nz', 'nx1', 'ny1', 'nxhalo', 'nyhalo'}
+    - units: Dict{'areaflux', 'courant', 'mass', 'wind'}
+    - origins: Dict{'compute_3d'}
+    - backend: 'numpy'
+    - fvt_dict: Dict{'grid_type', 'hord'}
+    - dt_acoustic: acoustic time step (default = 300 seconds)
+
+    Outputs:
+    - fluxPrep: Dict{'crx', 'cry', 'mfxd', 'mfyd', 'ucv', 'vcv', 'xaf', 'yaf'}
+        - crx, cry are Courant numbers
+        - ucv, vcv are the contravariant winds
+        - xaf, yaf are fluxes of area (dx*dy but incorporating wind speed)
+        - mfxd, mfyd are mass (volume) fluxes incorporating wind speed
+    - tracAdv: class instance that performs tracer advection
+    """
+
+    fluxPrep = run_finiteVolumeFluxPrep(
+        configuration,
+        uc,
+        vc,
+        delp,
+        density,
+        dimensions,
+        units,
+        origins,
+        backend,
+        dt_acoustic,
+    )
+
+    tracAdv = build_tracerAdvection(configuration, fvt_dict, tracers)
+
+    tracAdv_data = {
+        "tracers": tracers,
+        "delp": delp,
+        "mfxd": fluxPrep["mfxd"],
+        "mfyd": fluxPrep["mfyd"],
+        "crx": fluxPrep["crx"],
+        "cry": fluxPrep["cry"],
+    }
+
+    return tracAdv_data, tracAdv
+
+
+def run_advectionStepWithReset(
+    tracAdv_dataInit, tracAdv_data, tracAdv, dt, mpi_rank=None
```

Comment:
why is rank optional?

Reply:
If you pass it rank, it will write out more detailed diagnostics on rank 0. Like difference in tracer from one timestep to the next, and from initial condition. If not, no details get printed out